gms/gossiper.cc

```diff
@@ -853,34 +853,31 @@ future<semaphore_units<>> gossiper::lock_endpoint_update_semaphore() {
     return get_units(_endpoint_update_semaphore, 1, _abort_source);
 }
 
-future<> gossiper::mutate_live_and_unreachable_endpoints(std::function<void(gossiper&)> func) {
+future<> gossiper::mutate_live_and_unreachable_endpoints(std::function<void(live_and_unreachable_endpoints&)> func,
+        std::function<void(gossiper&)> on_success) {
     auto lock = co_await lock_endpoint_update_semaphore();
-    func(*this);
+    auto cloned = std::make_unique<live_and_unreachable_endpoints>(_live_endpoints, _unreachable_endpoints);
+    func(*cloned);
 
     // Bump the _live_endpoints_version unconditionally,
     // even if only _unreachable_endpoints changed.
     // It will trigger a new round in the failure_detector_loop
     // but that's not too bad as changing _unreachable_endpoints
     // is rare enough.
-    _live_endpoints_version++;
-    co_await replicate_live_endpoints_on_change();
+    co_await replicate_live_endpoints_on_change(make_foreign(std::move(cloned)), _live_endpoints_version + 1);
+
+    on_success(*this);
 }
 
 future<std::set<inet_address>> gossiper::get_live_members_synchronized() {
-    return container().invoke_on(0, [] (gms::gossiper& g) -> future<std::set<inet_address>> {
-        auto lock = co_await g.lock_endpoint_update_semaphore();
-        std::set<inet_address> live_members = g.get_live_members();
-        co_await g.replicate_live_endpoints_on_change();
-        co_return live_members;
+    return container().invoke_on(0, [] (gms::gossiper& g) {
+        return g.get_live_members();
```

Comment:
This might increase test flakiness that was previously solved by introducing the synchronized() functions.

Reply:
But even if this function grabs the semaphore and returns some snapshot of live members it may change right after it releases the semaphore, so why does it matter if it loses the race with `replicate_live_endpoints_on_change`?

With the exception safety in place any value we sample is bound to have been replicated to all shards at one point in time.

Reply:
Yes it might change immediately, but the returned value has been observed by other shards **before** that, which is the whole point.

IIRC the original scenario that motivated the `synchronized()` functions was something as follows:
a test was restarting some node X and then starting removenode operation on another node before all nodes noticed that X is up. In the middle of streaming nodes would notice that X is UP, triggering gossiper notifications, which interrupted streaming (`stream_manager::on_restart` closes existing streaming sessions). So the test was modified to wait for nodes to see that X is up before triggering removenode. However, that didn't help, because the live nodes endpoint returned data from one shard only (shard 0), and the problem still occured, because other shards would notice that the endpoint is dead after removenode was triggered and close streaming sessions... The solution was to ensure that all shards see the new alive endpoints before triggering removenode.

Reply:
> With the exception safety in place any value we sample is bound to have been replicated to all shards at one point in time.

Not have been, it might be still happening.

Reply:
The reason we needed the "synchronized" functions was that the state of live and unreachable endpoint was stale on non-zero shards AFAIK.  I think that if it can provide monotonic progression of the state it should be good enough for the api users.

Reply:
That said, I don't mind add an acquisition of the semaphore for extra safety.

Reply:
> The reason we needed the "synchronized" functions was that the state of live and unreachable endpoint was stale on non-zero shards AFAIK

Yes and this PR brings back that problem. The returned value might not have been replicated to non-zero shards yet i.e. state on non-zero shards might be stale.

Reply:
Followup PR: https://github.com/scylladb/scylladb/pull/16629